Objects/exceptions.c

```diff
@@ -1034,18 +1035,21 @@ BaseExceptionGroup_str(PyBaseExceptionGroupObject *self)
         self->msg, num_excs, num_excs > 1 ? "s" : "");
 }
 
+/*[clinic input]
+@critical_section
+BaseExceptionGroup.derive
+    excs: object
+    /
+[clinic start generated code]*/
+
 static PyObject *
-BaseExceptionGroup_derive(PyObject *self_, PyObject *excs)
+BaseExceptionGroup_derive_impl(PyBaseExceptionGroupObject *self,
+                               PyObject *excs)
+/*[clinic end generated code: output=4307564218dfbf06 input=f72009d38e98cec1]*/
 {
-    PyBaseExceptionGroupObject *self = _PyBaseExceptionGroupObject_cast(self_);
-    PyObject *init_args = PyTuple_Pack(2, self->msg, excs);
-    if (!init_args) {
-        return NULL;
-    }
-    PyObject *eg = PyObject_CallObject(
-        PyExc_BaseExceptionGroup, init_args);
-    Py_DECREF(init_args);
-    return eg;
+    PyObject *stack[] = {self->msg, excs};
+    size_t nargsf = 2 | PY_VECTORCALL_ARGUMENTS_OFFSET;
+    return PyObject_Vectorcall(PyExc_BaseExceptionGroup, stack, nargsf, NULL);
```

Comment:
This makes it thread safe?

Reply:
No, it was an unrelated optimization merged accidently into this, removed now.

Reply:
What here makes it thread safe then?

Reply:
> What here makes it thread safe then?

Adding `@critical_section` using argument clinic makes it use critical section around function call which makes it thread safe.